src/zarr/core/group.py

```diff
@@ -199,15 +524,87 @@ async def open(
             raise ValueError(f"unexpected zarr_format: {zarr_format}")
 
         if zarr_format == 2:
-            # V2 groups are comprised of a .zgroup and .zattrs objects
+            # this is checked above, asserting here for mypy
             assert zgroup_bytes is not None
-            zgroup = json.loads(zgroup_bytes.to_bytes())
-            zattrs = json.loads(zattrs_bytes.to_bytes()) if zattrs_bytes is not None else {}
-            group_metadata = {**zgroup, "attributes": zattrs}
+
+            if use_consolidated and maybe_consolidated_metadata_bytes is None:
+                # the user requested consolidated metadata, but it was missing
+                raise ValueError(consolidated_key)
+
+            elif use_consolidated is False:
+                # the user explicitly opted out of consolidated_metadata.
+                # Discard anything we might have read.
+                maybe_consolidated_metadata_bytes = None
+
+            return cls._from_bytes_v2(
+                store_path, zgroup_bytes, zattrs_bytes, maybe_consolidated_metadata_bytes
+            )
         else:
             # V3 groups are comprised of a zarr.json object
             assert zarr_json_bytes is not None
-            group_metadata = json.loads(zarr_json_bytes.to_bytes())
+            if not isinstance(use_consolidated, bool | None):
+                raise TypeError("use_consolidated must be a bool for Zarr V3.")
+
+            return cls._from_bytes_v3(
+                store_path,
+                zarr_json_bytes,
+                use_consolidated=use_consolidated,
+            )
+
+    @classmethod
+    def _from_bytes_v2(
+        cls,
+        store_path: StorePath,
+        zgroup_bytes: Buffer,
+        zattrs_bytes: Buffer | None,
+        consolidated_metadata_bytes: Buffer | None,
+    ) -> AsyncGroup:
+        # V2 groups are comprised of a .zgroup and .zattrs objects
+        zgroup = json.loads(zgroup_bytes.to_bytes())
+        zattrs = json.loads(zattrs_bytes.to_bytes()) if zattrs_bytes is not None else {}
+        group_metadata = {**zgroup, "attributes": zattrs}
+
+        if consolidated_metadata_bytes is not None:
+            v2_consolidated_metadata = json.loads(consolidated_metadata_bytes.to_bytes())
+            v2_consolidated_metadata = v2_consolidated_metadata["metadata"]
+            # We already read zattrs and zgroup. Should we ignore these?
+            v2_consolidated_metadata.pop(".zattrs")
+            v2_consolidated_metadata.pop(".zgroup")
```

Comment:
In the future, we may choose to inform the caller that the consolidated metadata does not match the root group metadata but for now, this seems fine.